python/cugraph-dgl/examples/graphsage/node-classification.py

```diff
@@ -0,0 +1,236 @@
+        prefetch_node_feats=["feat"],
+        prefetch_labels=["label"],
+    )
+    use_uva = args.mode == "mixed"
+    batch_size = 1024 * 20
+    train_dataloader = DataLoader(
+        g,
+        train_idx,
+        sampler,
+        device=device,
+        batch_size=batch_size,
+        shuffle=True,
+        drop_last=False,
+        num_workers=0,
+        use_uva=use_uva,
+    )
+
+    val_dataloader = DataLoader(
+        g,
+        val_idx,
+        sampler,
+        device=device,
+        batch_size=batch_size,
+        shuffle=True,
+        drop_last=False,
+        num_workers=0,
+        use_uva=use_uva,
+    )
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        st = time.time()
+        for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
+
+            x = blocks[0].srcdata["feat"]
+            y = blocks[-1].dstdata["label"]
+            y_hat = model(blocks, x)
+            loss = F.cross_entropy(y_hat, y)
+            opt.zero_grad()
+            loss.backward()
+            opt.step()
+            total_loss += loss.item()
+
+        et = time.time()
+
+        print(f"Time taken for epoch {epoch} with batch_size {batch_size} = {et-st} s")
+        acc = evaluate(model, g, val_dataloader)
+        print(
+            "Epoch {:05d} | Loss {:.4f} | Accuracy {:.4f} ".format(
+                epoch, total_loss / (it + 1), acc.item()
+            )
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--mode",
+        default="cugraph_storage",
+        choices=["cpu", "mixed", "puregpu", "cugraph_storage"],
+        help="Training mode."
+        " 'cpu' for CPU training,"
+        " 'mixed' for CPU-GPU mixed training, "
+        " 'puregpu' for CPU-GPU mixed training, "
+        " 'cugraph_storage' for pure-GPU training.",
+    )
+
+    parser.add_argument("--use_rmm", action="store_true")
+    args = parser.parse_args()
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # load and preprocess dataset
+    print("Loading data")
+    dataset = AsNodePredDataset(DglNodePropPredDataset("ogbn-products"))
+    g = dataset.g
+    if args.mode == "cugraph_storage":
+        if args.use_rmm:
+            import rmm
+
+            rmm.reinitialize(
+                pool_allocator=True, initial_pool_size=5e9, maximum_pool_size=25e9
+            )
+
+        g = cugraph_dgl.cugraph_storage_from_heterograph(g)
+        del dataset.g
+
+    else:
+        g = g.to("cuda" if args.mode == "puregpu" else "cpu")
+    device = torch.device("cpu" if args.mode == "cpu" else "cuda")
+
+    # create GraphSAGE model
+    feat_shape = (
+        g.get_node_storage(key="feat", ntype="_N")
+        .fetch(torch.LongTensor([0]).to(device), device=device)
+        .shape[1]
+    )
```

Comment:
why do not you first get the `shape[1]`, then use `to(device)` ? this will save some data movement time.

Reply:
Yes, this snippet works across all DGL storage devices  following the rfc at [link](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag#Extending-DGL-Graph-Sampling-Pipeline ) 

Also, see this example [link](https://github.com/dmlc/dgl/blob/2caef17eb8481ea42503a9843aad77af2710b425/python/dgl/storages/pytorch_tensor.py#LL28C47-L28C47)

> why do not you first get the shape[1], then use to(device) ? this will save some data movement time.

1. The `to(device)` call here is for the indices argument which needs the indices to be on the same device as the feature. 

2. We also are only moving 1 single index to device here which is an array of size 1 so i dont think this is expensive at all.